### PR TITLE
[GOVCMS-5606] Audit disallowed modules in ci

### PR DESCRIPTION
**Pull request by Ahmedjabar**
Adds a check for disallowed modules in CI.

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+version: '2.3'
+x-volumes:
+  &default-volumes
+    volumes:
+      - .:/app/
+volumes:
+  app: {}
+services:
+  test:
+    image: quay.io/govcms/govcms-ci
```
Comment (steveworley) [inline docker-compose.yml:11]:
```suggestion
    image: govcms/govcms-ci
```

File: tests/bats/validate/govcms-validate-modules.bats [hunk 2]
```diff
@@ -0,0 +1,40 @@
+#!/usr/bin/env bats
+# shellcheck disable=SC2002,SC2031,SC2030,SC2155
+
+load ../_helpers_govcms
+
+@test "No exported config files." {
+  export GOVCMS_CONFIG_FOLDER_PATH="tests/bats/validate/fixtures/config"
+  export GOVCMS_CORE_EXTENTION_FILE_NAME=""
+  export GOVCMS_DISALLOWED_MODULES="update
```
Comment (steveworley) [inline tests/bats/validate/govcms-validate-modules.bats:9]: I don't think we can do multiline variables in Lagoon like this - so we might need to use a delimiter and `ifs` to iterate there.

File: scripts/validate/govcms-validate-modules [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+FAILURES=""
+
+for file in $config_file; do
+  # Extract the list of enabled modules from core.extenions.yml:
+  ## the first sed command captures everything between `module:` and `theme:` keys in the file.
+  ## the returned result inclues `module:` and `theme:` so the second command removes the first and the last lines.
+  ## the final command removes tab indents.
+  enabled_modules=$(sed -n '/module:/,/theme:/p' "$file" | sed '1d;$d' | sed -e 's/[ \t]*//')
```
Comment (steveworley) [inline scripts/validate/govcms-validate-modules:35]: Awesome `sed` wondering if `yq` might simplfy this a little, the govcms-ci image has `yq` available.

```
yq r "$file" 'module'
```
That should generate a list like:
```
admin_toolbar: 0
ctools_block: 0
datetime: 0
datetime_range: 0
dblog: 0
ga_login: 0
views_ui: 0
webform: 0
webform_ui: 0
workflows: 0
ds: 1
menu_link_content: 1
pathauto: 1
ds_extras: 2
views: 10
paragraphs: 11
linked_field: 100
update: 0
govcms: 1000
```

Comment (steveworley) [reply scripts/validate/govcms-validate-modules:35]: We might be able to do 

```
yq r "$file" "module.$module_name"
```

Which will output the value or null eg.

```


That will output the value or null

```
module_name=paragraphs

yq r "$file" "module.$module_name"
11
```

File: scripts/validate/govcms-validate-modules [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  for module_name in $GOVCMS_DISALLOWED_MODULES; do
+    if [ $(is_module_enabled "$module_name" "$enabled_modules") -gt 0 ]; then
+       FAILURES="$FAILURES, $module_name"
+    fi
+  done
+
+done
+
+if [ -z "${FAILURES}" ]; then
+  echo "[success]: No disallowed module found in the exported config files."
```
Comment (steveworley) [inline scripts/validate/govcms-validate-modules:46]: Can we add the formatted output here too?

https://github.com/govCMS/scaffold-tooling/blob/develop/scripts/validate/govcms-prevent-theme-modules#L51

```
govcms-prepare-xml --failures="${FAILURES}" --total="${FILE_LFS}" --name="${GOVCMS_OUTFILE}" --fail-message="GovCMS.QA.FindModule"
```

We can set the message to `GovCMS.QA.AllowedModules` or something similar. This is a new feature to allow these scripts report in the junit format to allow for better integration into gitlab.

File: scripts/validate/govcms-validate-modules [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+# This will lint exported configuration files to identify if
+# any disallowed module has been enabled via exports.
+#
+
+GOVCMS_CONFIG_FOLDER_PATH=${GOVCMS_CONFIG_FOLDER_PATH:-"config"}
+GOVCMS_CORE_EXTENTION_FILE_NAME=${GOVCMS_CORE_EXTENTION_FILE_NAME:-"core.extension.yml"}
+GOVCMS_DISALLOWED_MODULES=${GOVCMS_DISALLOWED_MODULES:-}
+
+config_file=$(find "$GOVCMS_CONFIG_FOLDER_PATH" -type f \( -name "$GOVCMS_CORE_EXTENTION_FILE_NAME" \))
+
```
Comment (steveworley) [inline scripts/validate/govcms-validate-modules:17]:
```suggestion
echo "GovCMS Validate :: Verify enabled modules"
```

File: scripts/validate/govcms-validate-modules [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+#
+# GovCMS disallowed modules check.
+#
+# This will lint exported configuration files to identify if
+# any disallowed module has been enabled via exports.
+#
+
+GOVCMS_CONFIG_FOLDER_PATH=${GOVCMS_CONFIG_FOLDER_PATH:-"config"}
+GOVCMS_CORE_EXTENTION_FILE_NAME=${GOVCMS_CORE_EXTENTION_FILE_NAME:-"core.extension.yml"}
+GOVCMS_DISALLOWED_MODULES=${GOVCMS_DISALLOWED_MODULES:-}
```
Comment (steveworley) [inline scripts/validate/govcms-validate-modules:14]: Do we have a default list of modules that we want to ensure are disabled?

Comment (klonos) [reply scripts/validate/govcms-validate-modules:14]: dblog and update are the 2 that come to mind @steveworley

File: scripts/validate/govcms-validate-modules (hunk 3 above)
Comment (Ahmedjabar) [reply scripts/validate/govcms-validate-modules]: Awesome tip! I saw jq being used to parse json files but didn't notice yq was there as well. It significantly simplified the code.

File: scripts/validate/govcms-validate-modules (hunk 4 above)
Comment (Ahmedjabar) [reply scripts/validate/govcms-validate-modules:48]: Done.

File: scripts/validate/govcms-validate-modules (hunk 6 above)
Comment (Ahmedjabar) [reply scripts/validate/govcms-validate-modules]: Yeah @klonos mentioned, so far dblog and update are the main disallowed modules. I have made them the default value.

File: tests/bats/validate/govcms-validate-modules.bats (hunk 2 above)
Comment (Ahmedjabar) [reply tests/bats/validate/govcms-validate-modules.bats]: Done. 👍

**Review by steveworley — APPROVED** [review]
:shipit:

**Pull request merged**